### ISSUE TITLE
PendingDeprecationWarnings and ResourceWarnings with Python 3.2

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#810]
Set PYTHONWARNINGS="d" to enable warnings:

```
$ PYTHONPATH="build/lib" PYTHONWARNINGS="d" nosetests-3.2 -w tests
/usr/lib64/python3.2/site-packages/nose/core.py:247: ResourceWarning: unclosed file <_io.TextIOWrapper name='/usr/lib64/python3.2/site-packages/nose/usage.txt' mode='r' encoding='UTF-8'>
  os.path.dirname(__file__), 'usage.txt'), 'r').read()
........../usr/lib64/python3.2/site-packages/docutils/parsers/rst/states.py:2087: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/wrongenc.inc'>
  result = directive_instance.run()
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/bom.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:118: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  line = '<a name="L&#37;s"> </a>' &#37; no + escape(line)
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/insert_similar.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/images.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/extapi.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/contents.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/deleted_end.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/subdir/images.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/insert_beginning.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/autosummary.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/original.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/math.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/metadata.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/includes.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/objects.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/index.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/autodoc.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/deleted.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/added.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/modified.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/markup.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/extensions.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/doctest.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/generated/sphinx.application.Sphinx.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/versioning/insert.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/changes.py:131: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/subdir/includes.txt'>
  f = codecs.open(targetfn, 'w', 'latin1')
./tmp/Sphinx-1.1.2/build/lib/sphinx/highlighting.py:108: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  return '<pre>' + cgi.escape(source) + '</pre>\n'
./tmp/Sphinx-1.1.2/build/lib/sphinx/builders/htmlhelp.py:233: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  title = cgi.escape(node.astext()).replace('"','&quot;')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/htmlhelp.py:264: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  title = cgi.escape(title)
...../tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:240: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  title = escape(node.astext()).replace('"','&quot;')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:227: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  title = escape(refnode.astext()).replace('"','&quot;')
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:277: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  title = escape(title)
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:167: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  {'filename': escape(filename)})
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:182: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'outname': escape(outname),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:183: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'title': escape(self.config.html_title),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:184: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'version': escape(self.config.version),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:185: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'project': escape(self.config.project),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:186: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'namespace': escape(nspace),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:187: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'masterdoc': escape(self.config.master_doc),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:202: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'outname': escape(outname),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:203: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'title': escape(self.config.html_short_title),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:204: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'homepage': escape(homepage),
/tmp/Sphinx-1.1.2/build/lib/sphinx/builders/qthelp.py:205: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  'startpage': escape(startpage)})
....../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/autodoc.html'>
  tree.parse(source, parser)
.../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/extapi.html'>
  tree.parse(source, parser)
../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/markup.html'>
  tree.parse(source, parser)
......................................../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/includes.html'>
  tree.parse(source, parser)
............../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/objects.html'>
  tree.parse(source, parser)
......................./tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/contents.html'>
  tree.parse(source, parser)
.............../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/_static/statictmpl.html'>
  tree.parse(source, parser)
./tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/images.html'>
  tree.parse(source, parser)
..../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/bom.html'>
  tree.parse(source, parser)
./tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/subdir/includes.html'>
  tree.parse(source, parser)
...../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/subdir/images.html'>
  tree.parse(source, parser)
../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/extensions.html'>
  tree.parse(source, parser)
.../tmp/Sphinx-1.1.2/tests/etree13/ElementTree.py:1120: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/Sphinx-1.1.2/tests/root/_build/html/genindex.html'>
  tree.parse(source, parser)
.....info: not running latex, it doesn't seem to be installed
.........../usr/lib64/python3.2/site-packages/docutils/parsers/rst/states.py:2087: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/tmp01so5a/root/wrongenc.inc'>
  result = directive_instance.run()
.........................../usr/lib64/python3.2/site-packages/docutils/parsers/rst/states.py:2087: ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/tmpeqxzaz/root/wrongenc.inc'>
  result = directive_instance.run()
/usr/lib64/python3.2/site-packages/nose/case.py:198: DeprecationWarning: Please use assertEqual instead.
  self.test(*self.arg)
..................SS................./tmp/Sphinx-1.1.2/build/lib/sphinx/websupport/storage/differ.py:24: PendingDeprecationWarning: cgi.escape is deprecated, use html.escape instead
  proposal = escape(proposal)
.....
----------------------------------------------------------------------
Ran 221 tests in 73.302s

OK (SKIP=2)
```

(A minority of these warnings come from docutils and nose.)

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/810
- Originally reported by: Arfrever Frehtes Taifersar Arahesis
- Originally created at: 2011-11-02T01:32:12.954

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2012-01-29 11:23:26+00:00_

Closes #810: fix deprecation warnings with Python 3. What is left over is from nose or docutils.

→ <<cset 4ff9c71c91a0a6de0406ef8b8abbb8e96a84770b>>

**Comment by shimizukawa** [comment]
_From Arfrever Frehtes Taifersar Arahesis on 2013-03-06 20:34:35+00:00_

Actually 4 ResourceWarnings are caused by code in Sphinx, not nose or docutils. I attach the patch, which fixes them.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2013-03-06 20:44:41+00:00_

The intention is correct, but the patch is wrong: if "f" is never opened, the "finally" clause will raise an UnboundLocalError.

**Comment by shimizukawa** [comment]
_From Arfrever Frehtes Taifersar Arahesis on 2013-03-06 20:52:53+00:00_

I attach improved patch.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2013-03-07 06:19:41+00:00_

Closes #810: fix remaining PendingDeprecationWarnings with Python 3.

→ <<cset fea035ebe0e3ee4d143589300f32166a07d45919>>
